components/feature/media/src/main/java/mozilla/components/feature/media/service/MediaSessionServiceDelegate.kt

```diff
@@ -0,0 +1,162 @@
+import mozilla.components.feature.media.facts.emitStatePauseFact
+import mozilla.components.feature.media.facts.emitStatePlayFact
+import mozilla.components.feature.media.facts.emitStateStopFact
+import mozilla.components.feature.media.notification.MediaSessionNotification
+import mozilla.components.lib.state.ext.flowScoped
+import mozilla.components.support.base.ids.SharedIdsHelper
+import mozilla.components.support.base.log.logger.Logger
+import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifChanged
+
+/**
+ * Delegate handling callbacks from an [AbstractMediaSessionService].
+ *
+ * The implementation was moved from [AbstractMediaSessionService] to this delegate for better testability.
+ */
+internal class MediaSessionServiceDelegate(
+    private val context: Context,
+    private val service: AbstractMediaSessionService,
+    private val store: BrowserStore
+) {
+    private val logger = Logger("MediaSessionService")
+    private val notification by lazy { MediaSessionNotification(context, service::class.java) }
+    private val mediaSession: MediaSessionCompat by lazy { MediaSessionCompat(context, "MozacMediaSession") }
+    private var scope: CoroutineScope? = null
+    private var controller: MediaSession.Controller? = null
+
+    fun onCreate() {
+        logger.debug("Service created")
+        scope = store.flowScoped { flow ->
+            flow.ifChanged { it.tabs + it.customTabs }.map { state ->
+                state.tabs.filter { it.mediaSessionState != null } +
+                    state.customTabs.filter { it.mediaSessionState != null }
+            }.collect { states ->
+                processMediaSessionStates(states)
+            }
+        }
+    }
+
+    fun onDestroy() {
+        scope?.cancel()
+
+        logger.debug("Service destroyed")
+    }
+
+    fun onStartCommand(intent: Intent?) {
+        logger.debug("Command received: ${intent?.action}")
+
+        when (intent?.action) {
+            AbstractMediaSessionService.ACTION_LAUNCH -> {
+                // Nothing to do here. The service will subscribe to the store in onCreate() and
+                // update its state from the store.
+            }
+
+            AbstractMediaSessionService.ACTION_PLAY -> {
+                controller?.play()
+                emitNotificationPlayFact()
+            }
+            AbstractMediaSessionService.ACTION_PAUSE -> {
+                controller?.pause()
+                emitNotificationPauseFact()
+            }
+            else -> logger.debug("Can't process action: ${intent?.action}")
+        }
+    }
+
+    fun onTaskRemoved() {
+        /* no need to do this for custom tabs */
+        store.state.tabs.forEach {
+            it.mediaSessionState?.controller?.stop()
+        }
+
+        shutdown()
+    }
+
+    private fun processMediaSessionStates(sessionStates: List<SessionState>) {
+        val activeState = sessionStates.sortedByDescending { it.mediaSessionState }.firstOrNull()
```

Comment:
Is this sorting needed? Currently we assume there's only one state, right? All of this could probably already happen above when subscribing to the `Flow`, map it down to a single state (or null) and check whether it changed.

Reply:
It could be the way I'm subscribing.  I am seeing multiple `MediaSessionState`.  When I fix the subscription it could eliminate the need for sorting.  I'll investigate.  Thanks,

Reply:
I am still seeing multiple `SessionState` with `MediaSessionState` not null.